[권지인] sprint4

kjn4101 opened the pull request:
## 요구사항

### 기본
[o]  DiscodeitApplication의 테스트 로직은 삭제하세요.
[o]  지금까지 구현한 서비스 로직을 활용해 웹 API를 구현하세요.  
이때 @RequestMapping만 사용해 구현해보세요.

###API 테스트
[o] Postman을 활용해 컨트롤러를 테스트 하세요.
Postman API 테스트 결과를 다음과 같이 export하여 PR에 첨부해주세요.
[discodeit TestCate01.postman_collection.json](https://github.com/user-attachments/files/21749732/discodeit.TestCate01.postman_collection.json)


### 웹 API 요구사항
사용자 관리
[o] 사용자를 등록할 수 있다.
[o] 사용자 정보를 수정할 수 있다.
[o] 사용자를 삭제할 수 있다.
[o] 모든 사용자를 조회할 수 있다.
[o] 사용자의 온라인 상태를 업데이트할 수 있다.
권한 관리
[o] 사용자는 로그인할 수 있다.
채널 관리
[o] 공개 채널을 생성할 수 있다.
[o] 비공개 채널을 생성할 수 있다.
[o] 공개 채널의 정보를 수정할 수 있다.
[o] 채널을 삭제할 수 있다.
[o] 특정 사용자가 볼 수 있는 모든 채널 목록을 조회할 수 있다.
메시지 관리
[o] 메시지를 보낼 수 있다.
[o] 메시지를 수정할 수 있다.
[o] 메시지를 삭제할 수 있다.
[o] 특정 채널의 메시지 목록을 조회할 수 있다.
메시지 수신 정보 관리
[o] 특정 채널의 메시지 수신 정보를 생성할 수 있다.
[o] 특정 채널의 메시지 수신 정보를 수정할 수 있다.
[o] 특정 사용자의 메시지 수신 정보를 조회할 수 있다.
바이너리 파일 다운로드
[o] 바이너리 파일을 1개 또는 여러 개 조회할 수 있다.

[o]  웹 API의 예외를 전역으로 처리하세요.

### 심화
정적 리소스 서빙
[o]  사용자 목록 조회, BinaryContent 파일 조회 API를 다음의 조건을 만족하도록 수정하세요.
[o]  사용자 목록 조회
url: /api/user/findAll
요청
파라미터, 바디 없음
응답
ResponseEntity<List<UserDto>>
public record UserDto(
        UUID id,
        Instant createdAt,
        Instant updatedAt,
        String username,
        String email,
        UUID profileId,
        Boolean online
) {
}
[o]  BinaryContent 파일 조회
url: /api/binaryContent/find
요청
파라미터: binaryContentId
바디 없음
응답: ResponseEntity<BinaryContent>
[o]  다음의 파일을 활용하여 사용자 목록을 보여주는 화면을 서빙해보세요.

생성형 AI 활용
[o] 생성형 AI (Claude, ChatGPT 등)를 활용해서 위 이미지와 비슷한 화면을 생성 후 서빙해보세요.

## 주요 변경사항
- File*Repository 생성자 주입 문제 해결
- *Controller를 통한 웹 API 구현

## 스크린샷
<img width="1126" height="625" alt="스크린샷 2025-08-13 163733" src="https://github.com/user-attachments/assets/d73d0632-f082-4e1f-96a1-12c91679a561" />
<img width="1131" height="624" alt="스크린샷 2025-08-13 163920" src="https://github.com/user-attachments/assets/6d16aab9-bc34-41fd-b55c-0a0ca4662cb3" />


## 멘토에게
- UserStatus에서 Online 상태 업데이트 하는 부분에서 트러블이 많았습니다. 어떻게 손을 봐두긴 한 것 같은데 과제에서 요구한 사항이 이렇게 되는 게 맞는지는 잘 모르겠습니다. Access를 통해 유저가 Access한 시간이 300초 이내인 경우 Online으로 표시되는데, 딱히 채널을 생성하거나 메시지를 생성, 삭제, 수정 하는 등의 활동을 수행한다고 해서 Online으로 변하는 건 아니라서... 어떻게 관련 로직을 처리해야 하는지 잘 모르겠어서 일단 Access 한 경우에만 Online 처리 되도록 만들어 두었습니다. 
-

kjn4101 committed:
fix : File*Repository 생성자 주입 문제 해결

refactor : *Controller를 통해 웹 API 구현

lcy960729 commented on 5_sprint_mission_3/src/main/java/com/sprint/mission/discodeit/controller/AuthController.java:
```diff
@@ -0,0 +1,25 @@
+    @RequestMapping(value = "/login", method = RequestMethod.POST)
+    public ResponseEntity<UserResponseDto> login(@RequestBody UserLoginRequest userLoginRequest) {
+        UserResponseDto userResponse = authService.login(userLoginRequest);
+        return ResponseEntity.ok(userResponse);
```
로그인 실패도 처리되면 좋을것 같습니다. try catch를 이용해서 적절한 http 상태 코드를 반환해주면 될것 같네요. 401 (Unauthorized) 값을 반환하면 적절해보입니다!

lcy960729 commented on 5_sprint_mission_3/src/main/java/com/sprint/mission/discodeit/controller/BinaryContentController.java:
```diff
@@ -0,0 +1,32 @@
+    @RequestMapping(path = "find", method = RequestMethod.GET)
+    public ResponseEntity<Optional<BinaryContent>> find(@RequestParam("binaryContentId") UUID binaryContentId) {
+        Optional<BinaryContent> binaryContent = binaryContentService.findById(binaryContentId);
+        return ResponseEntity.status(HttpStatus.OK).body(binaryContent);
```
요구사항에는 Optional을 반환하는게 아니라 BinaryContent를 반환하는걸로 되어있습니다. 수정 부탁드리겠습니다.
추가로 없는 경우에는 404(Not Found) 상태코드를 반환해보는건 어떨까요??

lcy960729 commented on 5_sprint_mission_3/src/main/java/com/sprint/mission/discodeit/controller/ChannelController.java:
```diff
@@ -0,0 +1,84 @@
+
+    @RequestMapping(path = "public/create", method = RequestMethod.POST)
+    public ResponseEntity<Channel> createPublicChannel(@RequestBody PublicChannelCreateRequest publicChannelCreateRequest) {
+        Channel created = channelService.createPublicChannel(publicChannelCreateRequest);
```
단순히 created라는 변수명은 좋지 않은것 같습니다. newChannel과 같은 명확한 변수명을 사용하는 습관을 들이면 좋을것 같습니다.

lcy960729 commented on 5_sprint_mission_3/src/main/java/com/sprint/mission/discodeit/controller/ChannelController.java:
```diff
@@ -0,0 +1,84 @@
+    private final ChannelService channelService;
+
+    @RequestMapping(path = "public/create", method = RequestMethod.POST)
+    public ResponseEntity<Channel> createPublicChannel(@RequestBody PublicChannelCreateRequest publicChannelCreateRequest) {
```
Channel을 반환하는것이 아니라 Dto를 반환하도록 변경하면 api를 사용하는 클라이언트 입장에서 Channel과 ChannelDto 명세를 관리하는것이 아닌 하나의 모델에 대해서만 관리할 수 있습니다.

추가로 201 상태 코드를 반환할땐 response header로 location에 생성된 리소스 정보를 함께 내려줄수 있습니다.
location 헤더에 단일 채널을 조회하는 url를 함께 내려보내면 클라이언트는 location 헤더 값으로 채널을 조회하도록 할 수 있습니다.

lcy960729 commented on 5_sprint_mission_3/src/main/java/com/sprint/mission/discodeit/controller/ReadStatusController.java:
```diff
@@ -0,0 +1,50 @@
+                request.channelId(),
+                request.lastReadAt() != null ? request.lastReadAt() : Instant.now()
+        );
+        ReadStatus created = readStatusService.create(readStatus);
```
모델을 만드는 로직은 Service로 이동되면 좋겠습니다. 컨트롤러와 서비스 두곳에 로직이 분리된다면 로직 파악이 힘들뿐만 아니라 http api가 아닌 rpc 또는 메시지큐와 같은 presentation layer가 변경된다면 수정해야할 로직이 많아질겁니다!

lcy960729 commented on 5_sprint_mission_3/src/main/java/com/sprint/mission/discodeit/controller/ReadStatusController.java:
```diff
@@ -0,0 +1,50 @@
+        ReadStatus readStatus = readStatusService.findByUserIdAndChannelId(request.userId(), request.channelId())
+                .orElseThrow(() -> new IllegalArgumentException("ReadStatus가 존재하지 않습니다."));
+        readStatus.readUpdate(request.lastReadAt());
+        ReadStatus updated = readStatusService.update(readStatus);
```
해당 부분도 컨트롤러에 서비스 로직이 구현되어있네요 이동 부탁드립니다.

lcy960729 commented on 5_sprint_mission_3/src/main/java/com/sprint/mission/discodeit/entity/UserStatus.java:
```diff
@@ -29,16 +32,19 @@ public UserStatus(UUID userId, Instant lastAccessAt) {
     public boolean isOnline() {
-        return this.lastAccessAt != null && lastAccessAt.isAfter(Instant.now().minusSeconds(300));
+        return lastAccessAt != null &&
+                Instant.now().minusSeconds(300).isBefore(lastAccessAt);
```
online이라는 필드가 있음에도 isOnline이라는 함수가 있는 있을까요?? user에도 동일한 필드가 있는것으로 확인됩니다.

lcy960729 commented on 5_sprint_mission_3/src/main/java/com/sprint/mission/discodeit/service/basic/BasicChannelService.java:
```diff
@@ -108,27 +115,30 @@ public List<ChannelResponse> findAllByUserId(UUID userId) {
                     }
                 })
-                .toList();
+                .collect(Collectors.toList());
```
toList()에서 collect()로 변경한 이유가 있을까요?? toList()가 더 최신 문법이고 짧게 작성 가능합니다

lcy960729 reviewed (commented):
요구사항 잘 구현하셨고 일부 부족한 부분 리뷰 남겼습니다.

online 처리는 유저가 api를 요청할때 마다 업데이트 되도록 해보는건 어떨까요? 
물론 실제 서비스에서는 이렇게만 구현하긴 어려워 보이지만 현재 프로젝트에서는 가장 간단하게 구현 가능하다고 생각이 되네요! 추후에 배우게 될 ServletFilter나 Interceptor 라는것을 배우게 되면 쉽게 구현하실수 있을겁니다!

한달동안 고생하셨습니다!

lcy960729 commented on 5_sprint_mission_3/src/main/resources/static/script.js:
```diff
@@ -0,0 +1,65 @@
+async function fetchAndRenderUsers() {
+    // API 호출 대신 AI 데이터 사용
+    renderUserList(fakeUsers);
+}
```
실제 API를 사용하도록 변경해보면 좋을것 같습니다. Spring Controller에서 ModelAndView를 반환하여 구현해보면 어떨까요??

kjn4101 committed:
Refactor : AuthController에서 로그인 실패 시 401 에러 반환,
BinaryContentController에서 반환값 BinaryContent로 수정, 존재하지 않을 경우 404 에러 반환,
User와 UserStatus에서의 isOnline 로직 개선,
ChannelController의 변수값 newChannel로 수정,
Controller에 Service 로직이 구현되어 있는 다수 코드 Service로 이동

Feature: UserWebController 및 Thymeleaf 연동 추가, AI 생성 API가 아니라 ModelAndView를 반환하도록 수정
